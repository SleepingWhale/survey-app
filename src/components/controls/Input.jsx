import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';

const getInput = type => {
  return class Input extends Component {
    static propTypes = {
      id: PropTypes.string.isRequired,
      onChange: PropTypes.func.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
      question: PropTypes.string
    };

    static defaultProps = {
      question: ''
    };

    constructor(props) {
      super(props);
      this.input = createRef();
    }

    componentDidMount() {
      this.input.current.focus();
    }

    onChange = e => {
      const { onChange } = this.props;
      const { value } = e.target;
      onChange(value);
    };

    render() {
      const { value, question, id } = this.props;

      return (
        <div className="form-group">
          <label htmlFor={id}>{question}</label>
          <input
            ref={this.input}
            type={type}
            className="form-control"
            onChange={this.onChange}
            value={value}
            id={id}
            autoComplete="off"
          />
        </div>
      );
    }
  };
};

export const InputText = getInput('text');
export const InputNumber = getInput('number');
