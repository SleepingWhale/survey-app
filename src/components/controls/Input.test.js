import React from 'react';
import { mount } from 'enzyme';
import { InputText, InputNumber } from './Input';

describe('InputText', () => {
  it('should render correctly', () => {
    const onChangeHandler = jest.fn();
    const component = mount(
      <InputText
        onChange={onChangeHandler}
        value="text"
        id="id"
        question="question"
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should call handler', () => {
    const onChangeHandler = jest.fn();
    const content = 'hello world';
    const component = mount(
      <InputText
        onChange={onChangeHandler}
        value=""
        id="id"
        question="question"
      />
    );
    const input = component.find('input').at(0);

    input.instance().value = content;
    input.simulate('change');

    expect(onChangeHandler).toHaveBeenNthCalledWith(1, content);
  });
});

describe('InputNumber', () => {
  it('should render correctly', () => {
    const onChangeHandler = jest.fn();
    const component = mount(
      <InputNumber
        onChange={onChangeHandler}
        value="111"
        id="id"
        question="question"
      />
    );
    expect(component).toMatchSnapshot();
  });
});
