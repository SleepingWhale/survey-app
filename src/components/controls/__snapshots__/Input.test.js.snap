// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputNumber should render correctly 1`] = `
<Input
  id="id"
  onChange={[MockFunction]}
  question="question"
  value="111"
>
  <div
    className="form-group"
  >
    <label
      htmlFor="id"
    >
      question
    </label>
    <input
      autoComplete="off"
      className="form-control"
      id="id"
      onChange={[Function]}
      type="number"
      value="111"
    />
  </div>
</Input>
`;

exports[`InputText should render correctly 1`] = `
<Input
  id="id"
  onChange={[MockFunction]}
  question="question"
  value="text"
>
  <div
    className="form-group"
  >
    <label
      htmlFor="id"
    >
      question
    </label>
    <input
      autoComplete="off"
      className="form-control"
      id="id"
      onChange={[Function]}
      type="text"
      value="text"
    />
  </div>
</Input>
`;
